README.md
[![Maintenance](https://img.shields.io/badge/Maintained%3F-yes-green.svg)](https://GitHub.com/optum/cloudig/graphs/commit-activity)
![build](https://github.com/optum/cloudig/workflows/build/badge.svg?branch=master)
[![Github all releases](https://img.shields.io/github/downloads/optum/cloudig/total.svg)](https://GitHub.com/optum/cloudig/releases/)

[comment]: <> (<a href="https://cla-assistant.io/Optum/cloudig"><img src="https://cla-assistant.io/readme/badge/Optum/cloudig" alt="CLA assistant" /></a>)

[![made-with-Go](https://img.shields.io/badge/Made%20with-Go-1f425f.svg)](http://golang.org)

![cloudig](./logo.png)

cloudig, or Cloudigest, is a simple CLI tool for creating reports from various cloud sources with user-provided comments. It is written in Go and currently uses AWS APIs to generate reports based on five pillars — operational excellence, security, reliability, performance efficiency, and cost optimization. CLI allows users to provide comments for documenting exceptions and work in progress. This report can be useful in various use cases such as a quality gate step in IaC CI/CD pipeline, daily monitoring for operations to keep account status of multiple accounts in green status, single source for enterprise security/compliance teams to endorse the accounts, etc.

The currently supported cloud sources are:

- [AWS IAM Reflect](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles.html): This custom source uses [CloudTrail](https://aws.amazon.com/cloudtrail/) for event history of AWS IAM role activity and IAM to compare the actual permissions.

- [AWS Trusted Advisor](https://aws.amazon.com/premiumsupport/trustedadvisor/): Provides real time guidance to help provision resources following AWS best practices. 100+ checks on cost optimization, security, fault tolerance, performance, and service limits

- [AWS Config](https://aws.amazon.com/config/): Continuously monitors and records our AWS resource configurations and allows us to automate the evaluation of recorded configurations against desired configurations. This [module](https://github.com/Optum/aws_config) can help enable AWS Config with the desired configuration baseline.

- [AWS Inspector](https://aws.amazon.com/inspector/): Automatically assesses applications running in EC2 instances for exposure, vulnerabilities, and deviations from best practices. After performing an assessment, Amazon Inspector produces a detailed list of security findings prioritized by level of severity.

- [AWS ECR Scan](https://docs.aws.amazon.com/AmazonECR/latest/userguide/image-scanning.html) : Provides image scanning to identify software vulnerabilities in container images.

- [AWS Health Notifications](https://aws.amazon.com/premiumsupport/technology/personal-health-dashboard/): Provides relevant and timely information to help manage events in progress, and provides proactive notification to help you plan for scheduled activities.

### Demo
![demo](doc/demo.gif)


### Installation

`cloudig` can be installed two ways:

1. Downloading latest release binary:

```bash
# <ARCH> Darwin_arm64
# <VERSION> 0.1.0

 curl -LO https://github.com/Optum/cloudig/releases/download/v<VERSION>/cloudig_<VERSION>_<ARCH>.tar.gz  \
 &&  tar -xf cloudig_<VERSION>_<ARCH>.tar.gz cloudig \
 && chmod +x cloudig \
 && mv cloudig /usr/local/bin \
 && cloudig --help
```

2. Cloning repo and running `make build` **NOTE: must have Go installed**

```bash
git clone https://github.com/Optum/cloudig.git

cd cloudig

make build

./cloudig --help
```

#### CLI Verbs

`get` - Get reports. Get ready-made reports directly from sources like Trusted Advisor, ECR scan, etc.

`reflect` - Reflect on resources. Custom reports based on past usage and current configurations. Ex: Reflect on IAM role usage.

#### Global Flags

`--help`,`-h` : Generate help documentation

`--version` : Prints version of the CLI

`--rolearn`: (Optional) Takes the comma separated list of role ARN's. Provided role must have minimum permission needed to pull findings from AWS sources. Also, your parent account must have trust relationship so it can assume the role. When no role is provided, findings are only from the account associated with provided credentials

`--cfile`, `-c`: (Optional) YAML file to provide user comments for each finding. When this file is not provided, each finding is treated as a new finding

`--region`, `-r`: (Optional) AWS region to get results from. Default is us-east-1

`--output`, `-o`: (Optional) Output of the report. Options: json, table, and mdtable. Default is JSON

`--verbose`, `-v`: (Optional) set log level, use 0 to silence, 1 for critical, 2 for warning, 3 for informational, 4 for debugging and 5 for debugging with AWS debug logging (default 3)

#### IAM Reflect source specific flags

`--identity`, `-i`: (Optional) One or more IAM Identities (users, groups, and roles) ARNs separated by a comma [,]. Only role ARN is supported today

`--identity-tags`, `-t`: (Optional) Set of tags in form [key:value] separated by [,] to find the targeted IAM Identities. Only role ARNs is supported today. Ignored when --identity is provided

`--usage`, `-u` : (Optional) Reflect Identity usage data (default true, if --errors/-e is not explicitly provided)

`--errors`, `-e` : (Optional) Reflect Identity error data (default true, if --usage/-u is not explicitly provided)

`--caller-identity` : (Optional) Include caller identity with the report(default false)

`--absolute-time` : (Optional) Specify both the start and end times for the time filter in the form 'startTime-endTime' 'mm/dd/yyyy-mm/dd/yyy' ex: '10/25/2020-10/31/2020'

`--relative-time` : (Optional) Specify a time filter relative to the current time in days. Default 1 day. Ignored when absolute-time is provided

#### Health source specific flags

`--details`, `-d`: (Optional) Health event descriptions are often very long, thus, by default, it is shortened to three sentences. If this flag is added, then the entire description is printed out.

`--pastdays`: (Optional) Number of past days to get results from. Default is all health events that are open / upcoming.

#### Source specific examples:

- [IAM Reflect](doc/reflectiam.md)
- [TrustedAdvisor](doc/trustedadvisor.md)
- [AWS Config](doc/awsconfig.md)
- [Inspector](doc/inspector.md)
- [ECR Scan](doc/ecrscan.md)
- [AWS Health](doc/health.md)

#### Disclaimer

Even though the goal of "IAM reflect" is following the "Principle of least privilege" by keeping "Configured Permissions" equal to "Actual used Permissions", it is not always possible. This is mainly because some of the API actions for various AWS services are not tracked in CloudTrail, such as S3 & CloudWatch data events.
Also, in some cases, API actions are named slightly differently. Don't let the perfect be the enemy of the good. Use this wisely!
More info [1](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/cloudtrail-aws-service-specific-topics.html#cloudtrail-aws-service-specific-topics-list) & [2](https://docs.aws.amazon.com/awscloudtrail/latest/userguide/logging-management-and-data-events-with-cloudtrail.html#logging-data-events)

Reflect uses CloudTrail event data stored in S3 and currently assumes partitions in the form `${Trail log location}/${region}/${year}/${month}/${day}`.
It returns zero results if it doesn't find the necessary partitions. We should be able to support different partition schemes in future releases.

#### Comment file(cfile):

Comment file provides a way to pass in user comments to findings from various sources. By default, CLI looks for the file name 'comments.yaml' to parse the comments.
User can also provide a different location by using flag `--cfile` or `-c`

Below is the sample file:

```yaml
- accountid: "111111111111"
  ta-findings:
    - SECURITY-IAM_Use: "**EXCEPTION:** We use Federation and IAM roles to manage resources in AWS . No users/groups created in IAM"
  config-findings:
    - IAM_POLICY_BLACKLISTED_CHECK: "**EXCEPTION:** Removed the AdministratorAccess policy since the default AWS_*_Admins uses the policy. Future enhancement would be to create a Custom Rule that no other Role can use the AdministratorAccess policy besides the AWS_*_Admins"
  inspector-findings:
    - CIS_Operating_System_Security_Configuration_Benchmarks-1.0: "**EXCEPTION:*** Description here"
  health-findings:
    - AWS_RDS_SECURITY_NOTIFICATION: "**EXCEPTION:** Description here"
```

#### IAM permission requirements

Sample Policy needed to run cloudig and ability to use assume role to run report across multiple accounts:

```hcl
data "aws_iam_policy_document" "cloudig_policy" {
  statement {
    effect = "Allow"

    actions = [
      "support:*",
      "config:GetComplianceSummaryByConfigRule",
      "config:GetComplianceDetailsByConfigRule",
      "config:DescribeComplianceByConfigRule",
      "config:DescribeComplianceByResource",
      "config:DescribeConfigRules",
      "health:DescribeEvents",
      "health:DescribeEventDetails",
      "health:DescribeAffectedEntities",
      "inspector:GetAssessmentReport",
      "inspector:ListAssessmentRuns",
      "inspector:ListAssessmentTemplates",
      "inspector:DescribeAssessmentTemplates",
      "inspector:DescribeAssessmentTargets",
      "inspector:DescribeResourceGroups",
      "ec2:DescribeInstances",
      "ec2:DescribeImages",
      "ecr:DescribeRepositories",
      "ecr:ListImages",
      "ecr:DescribeImages",
      "ecr:DescribeImageScanFindings",
      "iam:ListRolePolicies",
      "iam:ListRoles",
      "iam:ListRoleTags",
      "iam:GetRolePolicy",
      "iam:ListAttachedRolePolicies",
      "iam:GetPolicyVersion",
      "sts:GetCallerIdentity",
      "s3:ListBucket",
      "s3:GetObject",
      "s3:PutObject",
      "s3:GetBucketLocation",
      "cloudtrail:DescribeTrails",
      "athena:GetQueryExecution",
      "athena:StartQueryExecution",
      "athena:ListTableMetadata",
      "athena:ListDatabases",
      "athena:GetTableMetadata",
      "athena:ListDataCatalogs",
      "athena:GetQueryResults",
      "glue:GetTables",
      "glue:GetDatabase",
      "glue:CreateTable",
      "glue:GetDatabases",
      "glue:GetTable",
    ]
    resources = ["*"]
  }
}

data "aws_iam_policy_document" "cloudig_assume_role" {
  statement {
    effect = "Allow"

    actions = [
      "sts:AssumeRole",
    ]

    principals {
      type        = "AWS"
      identifiers = ["arn:aws:iam::111111111111:role/AWS_111111111111_Read"]
    }
  }
}

resource "aws_iam_role" "cloudig_role" {
  name               = "cloudig"
  assume_role_policy = data.aws_iam_policy_document.cloudig_assume_role.json
}

resource "aws_iam_role_policy" "cloudig_policy" {
  name   = "cloudig"
  role   = aws_iam_role.cloudig_role.id
  policy = data.aws_iam_policy_document.cloudig_policy.json
}
```

**Note**: Support doesn't let you allow or deny access to individual actions. Therefore, the Action element of a policy is always set to support:_.
Similarly, Support & Inspector don't provide resource-level access, so the Resource element is always set to _. Need access to EC2 for getting AMI information

### Developer Notes

#### Build

cloudig uses the Go Modules support built into Go 1.11. Make sure installed go version is at least 1.11

`go build -o cloudig` OR `make build`

Build for different Target OS and Platform:

`env GOOS=<target-OS> GOARCH=<target-architecture> go build -o cloudig`

OR

`make compile`

- This will compile the 64-bit binaries for Linux, Windows, and Mac

#### Testing

Running the tests: `make test`

Validating the test coverage :

```bash
go test -v -coverprofile cover.out github.com/Optum/cloudig/pkg/cloudig

go tool cover -html=cover.out // opens a html output in default browser

```

**Testing methodology**:

cloudig tests are written using the [GoMock](https://github.com/golang/mock) mocking framework and table-driven tests.

A few of the reasons we chose to use GoMock are:

1. Allows us to write tests for functions that make API calls without actually making the call.

2. Integrates well with the standard library testing package: `testing`.

Examples of these tests can be found under the `pkg/aws` and `pkg/cloudig` folders

The mocks that were generated can be found under the `pkg/mocks` folder and the commands used to generate them can be found [here](/pkg/mocks/README.md)

#### Logging

This project uses a very simple but effective [logger library](https://github.com/kris-nova/logger) that allows us to set a different logging level with the flag `--verbose` or `-v`. Different levels are 1 for critical, 2 for warning, 3 for informational, 4 for debugging, and 5 for debugging with AWS debug logging with the default being 3 (informational).

- Files supporting commands in `package cmd` and controller logic in `package cloudig` are instrumented to log at a different level.
- AWS interactions in `package aws` are purposefully kept out of adding more logging as we can always use `-v5` to enable AWS SDK debug logging. All methods in AWS client is expected to promptly propagate the error back to the caller to avoid duplicate logging.


CHANGELOG.md
## v0.1.5 ( 9 November 2021)

* Add filtering to ecr output to not display if no findings found
* Condense ecr report so multiple tags are together for one image digest

## v0.1.4 ( 1 September 2021)

* Add fix for trusted advisor metadata coming in unexpected order

## v0.1.3 ( 9 August 2021)

* Add --exclude-regions option to health command
* Add --include-regions option to health command

## v0.1.2

Fix bug for trusted advisor where metadata doesn't come in expected order

## v0.1.1

Fix bux for multi-role trusted advisor

## v0.1.0

Initial release


cmd/get.go
package cmd

import (
	"fmt"
	"os"
	"strings"
	"time"

	awslocal "github.com/Optum/cloudig/pkg/aws"
	"github.com/Optum/cloudig/pkg/cloudig"

	"github.com/kris-nova/logger"
	"github.com/spf13/cobra"
)

var (
	commentsFile          string
	roleARN               string
	output                string
	region                string
	pastDays              string
	healthExcludeRegions  string
	healthIncludeRegions  string
	details               bool
	ecrImageTag           string
	identityARNs          string
	identityTags          string
	includeUsage          bool
	includeErrors         bool
	includeCallerIdentity bool
	absoluteTime          string
	relativeTime          int
)

// getCmd represents the get command
var getCmd = &cobra.Command{
	Use:       "get trustedadvisor/awsconfig/inspector/health/ecrscan",
	Short:     "Get report findings",
	ValidArgs: []string{"trustedadvisor", "awsconfig", "inspector", "health", "ecrscan"},
	Args:      cobra.OnlyValidArgs,
	Run: func(cmd *cobra.Command, args []string) {
		if len(args) < 1 {
			fmt.Println("missing subcommands")
			os.Exit(1)
		}
	},
}

// reflectCmd represents the reflect command
var reflectCmd = &cobra.Command{
	Use:       "reflect iam",
	Short:     "Reflect on IAM role permissions",
	ValidArgs: []string{"iam"},
	Args:      cobra.OnlyValidArgs,
	Run: func(cmd *cobra.Command, args []string) {
		if len(args) < 1 {
			fmt.Println("missing subcommands")
			os.Exit(1)
		}
	},
}

// trustedAdvisorCmd represents the get trustedadvisor command
var trustedAdvisorCmd = &cobra.Command{
	Use:     "trustedadvisor",
	Short:   "Get AWS Trusted Advisor report findings",
	Aliases: []string{"ta", "t"},

	Run: func(cmd *cobra.Command, args []string) {
		execute(&cloudig.TrustedAdvisorReport{})
	},
}

// awsConfigCmd represents the config command
var awsConfigCmd = &cobra.Command{
	Use:     "awsconfig",
	Short:   "Get AWS Config report findings",
	Aliases: []string{"ac", "a"},

	Run: func(cmd *cobra.Command, args []string) {
		execute(&cloudig.ConfigReport{})
	},
}

// inspectorCmd represents the inspector command
var inspectorCmd = &cobra.Command{
	Use:     "inspector",
	Short:   "Get AWS Inspector report findings",
	Aliases: []string{"inspect", "ins", "i"},

	Run: func(cmd *cobra.Command, args []string) {
		execute(&cloudig.InspectorReports{Helper: &cloudig.InspectorHelper{}})
	},
}

// healthCmd represents the get health command
var healthCmd = &cobra.Command{
	Use:     "health",
	Short:   "Get AWS Health notifications' details",
	Aliases: []string{"he", "h", "healthnotifications", "healthnotification"},

	Run: func(cmd *cobra.Command, args []string) {
		excludeRegionsArr := strings.Split(healthExcludeRegions, ",")
		if healthExcludeRegions == "" {
			excludeRegionsArr = []string{}
		}
		includeRegionsArr := strings.Split(healthIncludeRegions, ",")
		if healthIncludeRegions == "" {
			includeRegionsArr = []string{}
		}
		flags := struct {
			Details        bool
			PastDays       string
			ExcludeRegions []string
			IncludeRegions []string
		}{
			Details:        details,
			PastDays:       pastDays,
			ExcludeRegions: excludeRegionsArr,
			IncludeRegions: includeRegionsArr,
		}

		execute(&cloudig.HealthReport{
			Flags: flags,
		})
	},
}

// ecrScanCmd represents the ecrscan command
var ecrScanCmd = &cobra.Command{
	Use:     "ecrscan",
	Short:   "Get ECR Image Scan report findings",
	Aliases: []string{"scan", "sc", "s"},
	Run: func(cmd *cobra.Command, args []string) {
		execute(&cloudig.ImageScanReports{
			Flags: cloudig.ImageScanReportFlags{
				Tag:    ecrImageTag,
				Region: region,
			},
		})
	},
}

// iamCmd represents the reflect IAM command
var iamCmd = &cobra.Command{
	Use:     "iam",
	Short:   "Reflect on IAM Role permissions",
	Aliases: []string{"i", "iamrole"},

	Run: func(cmd *cobra.Command, args []string) {

		var roles []string
		if identityARNs != "" {
			roles = strings.Split(identityARNs, ",")
		}

		// convert string "k1:v1,k2:v2" to map[string]string{"k1":"v1","k2":"v2"}
		tags := make(map[string]string)
		for _, v := range strings.Split(identityTags, ",") {
			kv := strings.Split(v, ":")
			if len(kv) == 2 {
				tags[kv[0]] = kv[1]
			}
		}

		// if -u or -e is not provided both will be true, if one of them is provided then the other one is false
		if !includeUsage && !includeErrors {
			includeUsage = true
			includeErrors = true
		}

		// validate absolute-time
		if absoluteTime != "" {
			var startTime, endTime time.Time
			errorMessage := "--absolute-time is wrong. It should be in the form 'startTime-endTime' 'mm/dd/yyyy-mm/dd/yyy' ex: '10/25/2020-10/31/2020'"
			dates := strings.Split(absoluteTime, "-")
			if len(dates) != 2 {
				fmt.Println(errorMessage)
				os.Exit(1)
			}
			startDate, endDate := dates[0], dates[1]
			format := "01/02/2006" // mm/dd/yyyy format
			var err error
			// convert to time
			startTime, err = time.Parse(format, startDate)
			if err != nil {
				fmt.Println(errorMessage)
				os.Exit(1)
			}
			endTime, err = time.Parse(format, endDate)
			if err != nil {
				fmt.Println(errorMessage)
				os.Exit(1)
			}
			if endTime.Before(startTime) {
				fmt.Println(errorMessage)
				os.Exit(1)
			}
		}

		execute(&cloudig.ReflectReport{
			Flags: cloudig.NewReflectFlags(region, roles, tags, includeUsage, includeErrors, includeCallerIdentity, absoluteTime, relativeTime),
		})

	},
}

func init() {
	rootCmd.AddCommand(getCmd)
	rootCmd.AddCommand(reflectCmd)

	getCmd.AddCommand(trustedAdvisorCmd)
	getCmd.AddCommand(awsConfigCmd)
	getCmd.AddCommand(inspectorCmd)
	getCmd.AddCommand(healthCmd)
	getCmd.AddCommand(ecrScanCmd)
	reflectCmd.AddCommand(iamCmd)

	// Here you will define your flags and configuration settings.
	rootCmd.PersistentFlags().StringVarP(&commentsFile, "cfile", "c", "comments.yaml", "Comments file name")
	rootCmd.PersistentFlags().StringVar(&roleARN, "rolearn", "", "One or more role ARNs seperated by a comma [,]")
	rootCmd.PersistentFlags().StringVarP(&output, "output", "o", "json", "Output of report. Options: [json, table, mdtable]. Default output is JSON")
	rootCmd.PersistentFlags().StringVarP(&region, "region", "r", "us-east-1", "AWS region to get results from")
	rootCmd.PersistentFlags().IntVarP(&logger.Level, "verbose", "v", 3, "set log level, use 0 to silence, 1 for critical, 2 for warning, 3 for informational, 4 for debugging and 5 for debugging with AWS debug logging (default 3)")
	// this is CLI , so turning of timestamp
	logger.Timestamps = false
	// healthCmd specific flags
	healthCmd.PersistentFlags().BoolVarP(&details, "details", "d", false, "Flag to indicate level of printing for each notification (default false)")
	healthCmd.PersistentFlags().StringVar(&pastDays, "pastdays", "", "Number of past days to get results from")
	healthCmd.PersistentFlags().StringVar(&healthExcludeRegions, "exclude-regions", "", "Set of regions separated by [,] to exclude Health Notifications from. Takes precedence over include-regions")
	healthCmd.PersistentFlags().StringVar(&healthIncludeRegions, "include-regions", "", "Set of regions separated by [,] to include Health Notifications from. Use \"global\" as a region if wanting notifications affecting all regions. Ignored when exclude-regions is set")

	// ecrScanCmd specific flags
	ecrScanCmd.PersistentFlags().StringVar(&ecrImageTag, "tag", "", "Tag of ECR image(s) to report scan results.")

	// iamCmd specific flags
	iamCmd.PersistentFlags().StringVarP(&identityARNs, "identity", "i", "", "One or more IAM Identities (users, groups, and roles) ARNs separated by a comma [,].Only role ARNs is supported today")
	iamCmd.PersistentFlags().StringVarP(&identityTags, "identity-tags", "t", "", "Set of tags in form [key:value] separated by [,] to find the targeted IAM Identities. Only role ARN is supported today. Ignored when --identity is provided")
	// if -u or -e is not provided both will be true, if one of them is provided then the other one is false
	iamCmd.PersistentFlags().BoolVarP(&includeUsage, "usage", "u", false, "Reflect Identity usage data (default true, if --errors/-e is not explicitly provided)")
	iamCmd.PersistentFlags().BoolVarP(&includeErrors, "errors", "e", false, "Reflect Identity error data (default true, if --usage/-u is not explicitly provided)")
	iamCmd.PersistentFlags().BoolVar(&includeCallerIdentity, "caller-identity", false, "Include caller identity with the report(default false)")
	iamCmd.PersistentFlags().StringVar(&absoluteTime, "absolute-time", "", "Specify both the start and end times for the time filter in the form 'startTime-endTime' 'mm/dd/yyyy-mm/dd/yyy' ex: '10/25/2020-10/31/2020'")
	iamCmd.PersistentFlags().IntVar(&relativeTime, "relative-time", 1, "Specify a time filter relative to the current time in days. Default 1 day. Ignored when absolute-time is provided")
}

func execute(report cloudig.Report) {
	sess, err := awslocal.NewAuthenticatedSession(region)
	if err != nil {
		logger.Critical("error creating aws session: %v", err)
		os.Exit(1)
	}

	// example type should be "*cloudig.HealthReport", we are spliting the string to get "HealthReport"
	rType := strings.Split(fmt.Sprintf("%T", report), ".")[1]
	logger.Debug("all root level flags:\ncommentsFile: %s\nroleARN: %s\noutput: %s\nregion: %s\nlogLevel: %d\n", commentsFile, roleARN, output, region, logger.Level)

	if rType == "HealthReport" {
		logger.Debug("all health command flags:\ndetails: %t\npastDays: %s\n", details, pastDays)
	}
	if rType == "ReflectReport" {
		logger.Debug("all reflect command flags:\nidentityARNs: %s\nidentityTags: %s\nincludeUsage: %t\nincludeErrors: %t\nincludeCallerIdentity: %t\nabsoluteTime: %s\nrelativeTime: %d\n", identityARNs, identityTags, includeUsage, includeErrors, includeCallerIdentity, absoluteTime, relativeTime)
	}

	err = cloudig.ProcessReport(sess, report, output, commentsFile, roleARN)
	if err != nil {
		logger.Critical("error creating '%s': %v", rType, err)
	}
}
